drivers/hmis/app/models/hmis/user.rb

```diff
@@ -84,16 +86,25 @@ def lock_access!(opts = {})
     super opts.merge({ send_instructions: false })
   end
 
+  # A permissions base is an entity or entities that grant permissions on the given entity. This can be the entity
+  # itself in the case of projects, or can be another entity in the case of files, which are granted permissions through
+  # their client or their enrollment. A result of nil indicates that there is no permissions base for the given entity.
+  # If there is a permissions base, the result will be zero or more entities to use as a permissions base. If the result
+  # is no entities, it means that the entity has a permissions base, but no entities that act as that base are present.
+  # For example, if a client is granted permissions through enrollments but has no enrollments, it would return no
+  # entities.
   private def permissions_base_for_entity(entity)
-    return [entity.data_source, *entity.projects] if entity.is_a? Hmis::Hud::Client
-    return entity if entity.is_a? Hmis::Hud::Organization
-    return entity if entity.is_a? Hmis::Hud::Project
+    return unless entity.present?
 
-    if entity.is_a? Hmis::File
-      return [entity.client.data_source, *entity.client.projects] unless entity.enrollment.present?
+    if entity.is_a? Hmis::Hud::Client
+      return entity.projects_including_wip if entity.enrolled?
 
-      return entity.enrollment.project
+      return Hmis::Hud::Project.viewable_by(self)
```

Comment:
I also wanted to call out this change just to make sure you all agree with it. The logic here is that the permissions base for the client would be their projects if they are enrolled. If not enrolled, the base would be all projects the user can see, which would effectively mean that they have a permission on the client if they have that permission in any of the projects they can see. This is better than making it a special case in the `#{permission}_for?` method and achieves the same thing as far as I can tell.

I believe that works and will behave in the right way, but I wanted to call it out for a second opinion in case my thinking is incorrect here.

Reply:
I think that makes sense... I don't quite understand the returning of a user's projects, or how that would help determine if they can see the client that has no enrollments, but I agree, you can see the client if you have _some_ ability to see clients and the client has no enrollments.

Reply:
Hmmm I think it makes sense, but the `Project.viewable_by` and subsequent access lookups seem unnecessary.  We don't need to do all that just to determine if you have that permission globally.

> This is better than making it a special case in the #{permission}_for? method and achieves the same thing as far as I can tell.

Is the alternative to return this here:
```
return entity.projects_including_wip if entity.is_a? Hmis::Hud::Client
```
and then on line 70 add:
```
# If Client has no enrollments, allow the permission if the user has it anywhere
# (check if its an array rather than if the entity is a client, because File may return client-based project list too..)
entity_is_unenrolled = base_entities.is_a?(ActiveRecord::Relation) && base_entities.empty? && (entity.is_a?(Client) || entity.is_a?(File))
return send("#{permission}?") if entity_is_unenrolled
```

?

If so, I think some version of the latter special case is probably preferred because we skip the unnecessary query to `Project.viewable_by(self)` which can be expensive. We need to build the project list, since we really just want to do a global check. Right?

Reply:
Yeah, that's the rub for me. A global permission lookup is preferable, but I don't really want to add a special condition at line 70, as I think that makes that method harder to reason about. I think I'll add another method called `uses_global_permissions?(entity)` to determine whether the entity should use global permissions or not, then add the special cases to that. I think that's better. Do you agree?